[Race Detector] data race in `pkg/hubble.TestRingReader_NextFollow_WithEmptyRing`

https://app.travis-ci.com/github/cilium/cilium/jobs/534366323

```
=== RUN   TestRingReader_NextFollow_WithEmptyRing
==================
WARNING: DATA RACE
Write at 0x00c0000958f8 by goroutine 56:
  runtime.racewrite()
      <autogenerated>:1 +0x24
  github.com/cilium/cilium/pkg/hubble/container.(*RingReader).Close()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader.go:117 +0x45
  github.com/cilium/cilium/pkg/hubble/container.TestRingReader_NextFollow_WithEmptyRing()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader_test.go:305 +0x652
  testing.tRunner()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1259 +0x22f
  testing.(*T).RunÂ·dwrapÂ·21()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1306 +0x47

Previous read at 0x00c0000958f8 by goroutine 57:
  runtime.raceread()
      <autogenerated>:1 +0x24
  github.com/cilium/cilium/pkg/hubble/container.(*RingReader).NextFollow()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader.go:81 +0x1bc
  github.com/cilium/cilium/pkg/hubble/container.TestRingReader_NextFollow_WithEmptyRing.func1()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader_test.go:295 +0x84

Goroutine 56 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1504 +0x9d1
  main.main()
      _testmain.go:303 +0x32b

Goroutine 57 (running) created at:
  github.com/cilium/cilium/pkg/hubble/container.TestRingReader_NextFollow_WithEmptyRing()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader_test.go:292 +0x576
  testing.tRunner()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1259 +0x22f
  testing.(*T).RunÂ·dwrapÂ·21()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1306 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x00c0000958e0 by goroutine 52:
  github.com/cilium/cilium/pkg/hubble/container.(*RingReader).NextFollow.func1()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader.go:86 +0x10e
  github.com/cilium/cilium/pkg/hubble/container.(*RingReader).NextFollowÂ·dwrapÂ·1()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader.go:89 +0x58

Previous read at 0x00c0000958e0 by goroutine 57:
  github.com/cilium/cilium/pkg/hubble/container.(*RingReader).NextFollow()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader.go:99 +0x40e
  github.com/cilium/cilium/pkg/hubble/container.TestRingReader_NextFollow_WithEmptyRing.func1()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader_test.go:295 +0x84

Goroutine 52 (running) created at:
  github.com/cilium/cilium/pkg/hubble/container.(*RingReader).NextFollow()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader.go:82 +0x2f8
  github.com/cilium/cilium/pkg/hubble/container.TestRingReader_NextFollow_WithEmptyRing.func1()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader_test.go:295 +0x84

Goroutine 57 (finished) created at:
  github.com/cilium/cilium/pkg/hubble/container.TestRingReader_NextFollow_WithEmptyRing()
      /home/travis/gopath/src/github.com/cilium/cilium/pkg/hubble/container/ring_reader_test.go:292 +0x576
  testing.tRunner()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1259 +0x22f
  testing.(*T).RunÂ·dwrapÂ·21()
      /home/travis/.gimme/versions/go1.17.linux.amd64/src/testing/testing.go:1306 +0x47
==================
    testing.go:1152: race detected during execution of test
--- FAIL: TestRingReader_NextFollow_WithEmptyRing (0.10s)
```

--- a/pkg/hubble/container/ring_reader.go
+++ b/pkg/hubble/container/ring_reader.go
@@ -8,6 +8,7 @@ import (
 	"sync"
 
 	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/cilium/cilium/pkg/lock"
 )
 
 // RingReader is a reader for a Ring container.
@@ -15,6 +16,7 @@ type RingReader struct {
 	ring          *Ring
 	idx           uint64
 	ctx           context.Context
+	mutex         lock.Mutex // protects writes to followChan
 	followChan    chan *v1.Event
 	followChanLen int
 	wg            sync.WaitGroup
@@ -75,16 +77,21 @@ func (r *RingReader) NextFollow(ctx context.Context) *v1.Event {
 	// if the context changed between invocations, we also have to restart
 	// readFrom, as the old readFrom instance will be using the old context.
 	if r.ctx != ctx {
+		r.mutex.Lock()
 		if r.followChan == nil {
 			r.followChan = make(chan *v1.Event, r.followChanLen)
 		}
+		r.mutex.Unlock()
+
 		r.wg.Add(1)
 		go func(ctx context.Context) {
 			r.ring.readFrom(ctx, r.idx, r.followChan)
+			r.mutex.Lock()
 			if ctx.Err() != nil && r.followChan != nil { // context is done
 				close(r.followChan)
 				r.followChan = nil
 			}
+			r.mutex.Unlock()
 			r.wg.Done()
 		}(ctx)
 		r.ctx = ctx
@@ -95,8 +102,12 @@ func (r *RingReader) NextFollow(ctx context.Context) *v1.Event {
 		}
 	}()
 
+	r.mutex.Lock()
+	followChan := r.followChan
+	r.mutex.Unlock()
+
 	select {
-	case e, ok := <-r.followChan:
+	case e, ok := <-followChan:
 		if !ok {
 			// the channel is closed so the context is done
 			return nil

--- a/pkg/hubble/container/ring_reader.go
+++ b/pkg/hubble/container/ring_reader.go
@@ -121,9 +121,10 @@ func (r *RingReader) NextFollow(ctx context.Context) *v1.Event {
 	}
 }
 
-// Close waits for any method to return and closes the RingReader. It is not
+// Close waits for any spawned go routines to finish. It is not
 // required to call Close on a RingReader but it may be useful for specific
-// situations such as testing.
+// situations such as testing. Must not be called concurrently with NextFollow,
+// as otherwise NextFollow spawns new go routines that are not waited on.
 func (r *RingReader) Close() error {
 	r.wg.Wait()
 	return nil

--- a/pkg/hubble/container/ring_reader_test.go
+++ b/pkg/hubble/container/ring_reader_test.go
@@ -289,11 +289,13 @@ func TestRingReader_NextFollow_WithEmptyRing(t *testing.T) {
 	reader := NewRingReader(ring, ring.LastWriteParallel())
 	ctx, cancel := context.WithCancel(context.Background())
 	c := make(chan *v1.Event)
+	done := make(chan struct{})
 	go func() {
 		select {
 		case <-ctx.Done():
 		case c <- reader.NextFollow(ctx):
 		}
+		close(done)
 	}()
 	select {
 	case <-c:
@@ -302,5 +304,6 @@ func TestRingReader_NextFollow_WithEmptyRing(t *testing.T) {
 		// the call blocked, we're good
 	}
 	cancel()
+	<-done
 	assert.Nil(t, reader.Close())
 }

